### ISSUE TITLE
Spell Resistance not correctly increasing the miss (full resist) chance of DoT effects (Gamebreaking)

### ISSUE THREAD

**Issue by Skuzzlez** [SunstriderEmu/BugTracker#3638]
**Describe the bug**
Increasing Spell Resistance does not correctly increase the full resist (miss) chance of damage over time spells and effects. This seems to also apply to other non-binary spells such as Mind Blast, SW:Death etc.

**To Reproduce**
Get a character / pet with high shadow resistance (ideally buffed with Shadow Protection). Binary spells such as Psychic Scream will often resist, however Damage over Time effects such as Shadow Word: Pain or Corruption seem not to resist at all, and only partially on individual ticks of the dot effect. This is not Blizz-like (as of patch 2.0), and hugely influences gameplay and the strengths / weaknesses of certain classes.

**Expected behavior**
After the 2.0 patch, DoT effects were changed to have the same resist chance as normal spells. Damage over time effects should be able to be fully resisted on the initial cast (no application), in addition to partially resisted on individual damage ticks.

Post from Kalgan (lead designer) after 2.0 patch and DoT nerf:

"Previously, dots in general were 1/10th as likely to be resisted as normal spells. That has been fixed, they're now as likely to be resisted as other spells. Please let us know whether there's any reason to believe they aren't working this way"

https://web.archive.org/web/20080601184008/http://forums.worldofwarcraft.com/thread.html?topicId=65457765&pageNo=18&sid=1#348

**Additional context**
As mentioned above, this HUGELY affects gameplay on this server.

**Comment by Skuzzlez** [comment]
Additionally, Frostbolt is incorrectly being considered a binary spell and cannot be partially resisted on this server. There is a thread here suggesting that this was changed in patch 2.3 so that Frostbolt is considered non-binary for the purpose of resisting spells:

https://github.com/cmangos/issues/issues/1193

Regarding DoTs, perhaps they are incorrectly being considered on this server as non-binary and not capable of full resists. The above thread and linked documentation suggests that they should in fact be considered binary for the purpose of being able to fully resist them.

**Comment by Jerrythemagicdog** [comment]
About Frostbolt:
the thread you linked from cmangos contradicts that later and links the same proof we have linked in the code:
```
    // Also from EJ: http://web.archive.org/web/20080811094101/http://elitistjerks.com/f31/t18441-mage_sweet_informational_thread/
    // "Frostbolt, however, is a binary spell, meaning it has special rules regarding magic resist. A binary spell can only ever hit or miss; 
    // it will not ever partially resist. Because of this, frostbolt is completely unaffected by level based magic resist. 
    // We're not exactly sure why this occurs, but frost mages almost always see a 99% hit rate on frostbolt (assuming they are hit capped 
    // like they should be)  Therefore the 6% damage reduction can be completely ignored. This does not mean that frostbolt does more damage 
    // than fireball; it is only another consideration that has been made when comparing specs and theorycrafting."
```

**Comment by Skuzzlez** [comment]
My mistake - you seem to be correct regarding Frostbolt. It seems it was changed in the WotLK PTR.

Regarding DoTs however, it still seems that the application of dots itself should be Binary and thus able to be fully resisted - with the individual damaging ticks of DoTs being considered Non-Binary. The classification of DoTs on this server may also explain why the initial cast of a damage over time effect on a hunter with Aspect of the Cheetah will (incorrectly) proc the daze from Aspect of the Cheetah, despite not having done any damage yet.

**Comment by Jerrythemagicdog** [comment]
Is there any proof of DoTs working that way? Any video of Warlocks dotting other Warlocks' pets during TBC (which is a bad sample size already due to the scarcity of that happening) shows no resists of DoT applications (quite the opposite), no mention of that either.

**Comment by Skuzzlez** [comment]
As someone explains in the other thread, in regards to how DoTs are classified:

DoTs are, in fact, just a component of a spell, an aura one. If there is no direct damage components within the spell, it ends up being binary. So "pure DoTs" are always binary on application. If there is a damage component in it next to a DoT (Fireball), the spell follow non-binary rules on application.

There is no such thing as "DoT spells". Internally, its either binary or non-binary spell with a DoT component in it. Affliction spells, on the other hand, are almost all exclusively binary, which is what they meant by "DoT spells".

https://wowwiki.fandom.com/wiki/Binary_spell
"DoTs used to be binary spells, but this was changed." This doesn't tell us much, but it confirms that DoTs were at one point in the game's life, indeed binary.

Blizzard definition of Binary / Non-Binary in May 2008:
https://web.archive.org/web/20080512210527/https://www.worldofwarcraft.com/info/basics/resistances.html

This suggests that spells that are primarily partially resisted are "Direct-Damage Spells", whereas those that can be fully resisted are Binary. From the fact that they specifically declare "Direct-Damage Spells" are the ones that are typically partially resisted, it can be inferred that DoT applications should be considered Binary, as they are not direct damage spells. Why would Blizzard make that distinction otherwise? This suggests that, at the very least, DoTs were considered Binary up until roughly this point (5 months before WotLK release).

Videos:

Here's a video of Veex, recorded during retail S2 - there's a few times throughout the video where he gets full resists on DoTs. Fortunately he has an addon as well which denotes when he gets resists. This unfortunately doesn't show what buffs the opponents had on, nor does it distinguish between 'resists' and 'misses', but it is highly unlikely that a R1 affliction warlock, of all people, would not be hit-capped.

https://youtu.be/86ZCLm910uo?t=826
13:51 - Siphon Life fully resists on the rogue.
15:21 - Unstable Affliction fully resists on the priest. (buffs unknown)
18:55 - Curse of Agony fully resists on the priest. (buffs unknown)


Pyrilus R1 video:
https://youtu.be/ao6sbIKoh3o?t=1400
10:41 - Corruption resists against a Felhunter with Mark of the Wild.
23:20 - Corruption Resists on a Priest with Shadow Protection (you can see his target in the lower half of the screen)

Not a huge sample size unfortunately, and it's difficult to know if they were truly hit capped or not, but it wouldn't be unreasonable to assume so.

**Comment by plebeianx** [comment]
https://youtu.be/yU6T3RA77ZE?t=1377

**Comment by Jerrythemagicdog** [comment]
Thanks for the video, the Warlock does put Curse of Shadows (later Curse of Elements) on the pet here so that significantly reduces the resist chance, not very conclusive, he's also definitely spell penetration capped for the Soul Link Warlock's resistances.

**Comment by kelno** [comment]
I found a big issue indeed here:
When casting a non binary spell, the hit chance should reduced by the 100% part of resist chance from spell resistances.
Blahblah this wasn't working properly and the chance was never changed there, so full resists from spell resistance could never happen on non binary spells (like dots).
Fixed with next update

**Comment by Skuzzlez** [comment]
Haven't noticed any changes as of the time of this post, though I'm not sure if the server has been restarted yet.

**Comment by kelno** [comment]
It wasn't applied yet.
Though after more analysis of this, this isn't as game breaking as anticipated, as the "full resist chance" doesn't move much unless you have extremely high level of resistance (+50%).
Another real issue that's been found with this though is that extra spell hit could be used to overcome the resist chance from spell resistance. That shouldn't be the case and has been fixed.

**Comment by Skuzzlez** [comment]
It's huge for protecting against DoT effects on pets. As argued above, DoT's should not be considered "direct-damage spells" (they're completely the opposite) and should be subject to the full resist chance. Blizzard clearly makes this distinction when they use the term "direct-damage spells"

They use this terminology in both the link you provided, and on their official website archived from 2008-2009:

https://web.archive.org/web/20080512210527/https://www.worldofwarcraft.com/info/basics/resistances.html

From WoW Wiki:

> Direct Damage, or DD, usually refers to any type of magic spell (or a component of a spell) whose direct effect is to cause burst damage (hitpoint loss) to one or multiple targets.
> 
> Examples
> Mage spell: Frostbolt
> Druid spell: Wrath
> Priest spell: Smite
> Warlock spell: Shadow Bolt
> Paladin spell: Exorcism
> Shaman spell: Lightning Bolt

https://wowwiki.fandom.com/wiki/Direct_damage

Therefore, damage over time effects should not be subject to the reduced resist chance that direct-damage spells are. They should be considered part of the "most spells" category and should have an increased resist chance. This is absolutely HUGE for high resistance units (i.e. paladins and pets, which are currently being absolutely melted by warlocks). This was how it was on retail TBC, and has been in every other TBC environment.

**Comment by Skuzzlez** [comment]
Here you go, somewhat definitive proof that DoTs should be binary until WotLK:

https://vanilla-wow.fandom.com/wiki/Resistance#:~:text=DoT%20spells%20are%20now%20considered,is%20applied%20on%20each%20tick.&text=If%20the%20probability%20is%20less%20than%200%2C%20it%20cannot%20occur.

> As of Wrath of the Lich King, the possible percentages of damage resisted are weighted in sections of 0%, 10%, 20%, ..., 100% averaging to the same overall damage mitigation.[3] However only the closest three or four sections will ever occur at a given resistance value. This allows a sufficiently high resistance to guarantee a certain resist percentage, even in the worst case. DoT spells are now considered non-binary and the partial resistance calculation is applied on each tick.

There are also a number of threads around 2010 / 2011 (after the release of WotLK) discussing this change, but nothing before:

https://lmgtfy.com/?q=%22dot+spells+are+now+considered+non-binary%22

Hope this settles the issue.

**Comment by kelno** [comment]
Okay I've been re reading those cmangos thread, this is really complicated stuff. Our implementation is based partly on cmangos implementation.
So indeed I see dots can be partially resisted... The sentence from wowwiki "DoT spells are now considered non-binary and the partial resistance calculation is applied on each tick." doesn't make sense then, since that second part was already true since vanilla. I think the WoWWiki page has some confusion there, probably some dots were changed from binary to non binary. (Not the first time WoWWiki is wrong or not precise either, from our own experience + pointed out by developers in the cmangos thread as well).

The "1/10th" seem to be a sonsensical thing blizzard had for classic and was then removed, we can't make much of that. cmangos is only trying to make sense of it because they want to have the blizzlike rule for their vanilla core.

I read something interesting from that cmangos thread:
https://github.com/cmangos/issues/issues/1193#issuecomment-279419101
"There is no such thing as "DoT spells". Internally, its either binary or non-binary spell with a DoT component in it. Affliction spells, on the other hand, are almost all exclusively binary, which is what they meant by "DoT spells". "
That could make sense, maybe we just need to reclassify some specific dots to binary. It could explain perceived higher resist chance on application for some dots.

You're suggesting here that dots should be binary but still have partial resists, that really goes against the definition of what's a binary spell. The very name "binary" is about resisting completely, or not at all. Plus it makes sense for gameplay as well, you get your spell resistance value somewhere, either on application, or on partial resists. Having it on both would mean spell resistance has twice the effectiveness for dots which makes no sense.
If anything I find spell resistance stronger if the resistance applies as partial resists (= dot is non binary), because that means the dot is easier to apply but does way less damage. Smoothed damage is easier to handle.

Okay so:
- On this core, some spells might be misclassified as non binary when they should be. It's hard to find data for specific spells but it really doesn't matter too much so I'll leave it at that.
- After review now that I have that whole subject in my head again, the rules implemented in the core seem to be working as intended, you do get the advertised value out of spell resistance.

**Comment by Skuzzlez** [comment]
Hi Kelno,

Regarding the effectiveness of dots being binary vs non-binary, you are correct in your assessment that under normal circumstances, they are equally as effective, in a PvE setting - sure. In a PvE setting this is irrelevant, but in a PvP setting global cooldowns are a precious resource. 

The ability to always apply dots and have them do a fixed amount of damage is a huge buff to say, warlocks against hunter teams for example. In previous TBC environments, it's usually not worth it for the warlock to spend global cooldowns trying to dot the pet since you can get upwards of 50% resist chance. Having affliction dots as non-binary completely negates this and makes every single one of their global cooldowns effective, and thus increases the amount of pressure they can do.

For example - currently a warlock can now reliably dot a hunter, his arena partner, and the pet (for 50% damage let's say), meaning their total DPS output is 250% assuming no resistances on the players. With DoTs as binary, the warlock has to spend much more time attempting to cast DoT spells on the pet - meanwhile his other DoTs are running out while doing this, and his overall damage is reduced because of the need to spend more global cooldowns.  It makes the pet not a reliable target to DoT (which is how it should be). In PvP situations, global cooldowns are a very precious resource.

I would strongly suggest that this is changed (fixed), as it provides a huge boost to all DoT classes against hunters (and warlocks to an extent), isn't Blizzlike, and is certainly one of the reasons why you don't see many high rated hunter teams on this server. Warlocks can reliably apply pressure to the pet while dotting 2+ other targets, it becomes unhealable.

**Comment by Skuzzlez** [comment]
Regardless, despite it seeming not logical, DoTs were both binary (fully resistible) and partially resistible up until WotLK. This is primarily because they had large spell damage coefficients and low cast times.

1. Application of DoTs is binary and fully resistible by resistance up until WotLK - established above

2. Damage component of DoTs is also partially resistible - established by the videos from Vanilla above. In Vanilla, DoTs have 1/10 the chance to partially resist as other damaging spells, confirmed by the thread below. It is unclear whether this refers only to partial resistance, or to full resistance too.
> https://web.archive.org/web/20080601184008/http://forums.worldofwarcraft.com/thread.html?topicId=65457765&pageNo=18&sid=1#348

3. In Vanilla & TBC, partial resistance damage was consistent for each tick, i.e. If the first tick resisted at 25%, each subsequent tick would resist at 25%. Other possibly resistance values were 50% or 75%.

4. In the 2.0 patch, DoTs were _still_ considered too strong, and had their resistance chance nerfed: "Previously, dots in general were 1/10th as likely to be resisted as normal spells. That has been fixed, they're now as likely to be resisted as other spells", established by the above thread. _This is unclear as to whether it refers to partial resists or full resists, but the lack of distinction implies both._ The language used by the original poster (whom Kalgan responded to) was the following:

> Also,
> 
> There are now reports that affliction spells (dots, fear) ie. the things that MAKE the warlock class are getting HEAVILY more resisted

Again, no distinction - but usually when someone refers to resist, they mean full resists and not partial ones. In my opinion, the lack of distinction implies both (along with memory of past experience).

TL:DR: 
DoTs were binary and both fully & partially resistible in Vanilla (albeit resistible at 1/10th of the rate of other spells). They were _still_ considered too powerful and were given the full chance to be resisted (seemingly full resist, and partial resist) when 2.0 hit. They weren't considered to be Non-Binary until WotLK.

**Comment by kelno** [comment]
I agree this makes a difference with the GCD in a PvP setting + The Nightfall case.

About point 3:
Your first video shows corruption ticks with various degrees of partial resistance, any idea why? https://www.youtube.com/watch?v=SRXUTgrZOvM&feature=youtu.be&t=16m55s

4: This is largely discussed in the cmangos thread as well and they conclude that change was about partial resists only (using more more data than we do here so I'd rather trust them until we have some)
Overall this says dots should have as much partial resists as other spells, but doesn't help me conclude that it should be considered as a binary spell on application and have high resist chance there as well.

That 1/10 thing definitely doesn't say anything conclusive to me, it says dots are going to be resisted like other spells, which doesn't point to a special rule applying to them, it rather feels like the removal of a bug where dots were incorrectly handled differently. The lack of clarity and detail of that post is also making this not a super strong source for deduce things.

If we're sure some spells should really resist more on application (which I still don't have data for, so I don't know), this could still be explained by said spells having been switched to binary at some point (with no more partial resists then). Which is a more likely explanation to me than having a special rule only for dots.
I know it's hard to find but what could help right now is TBC data (after that 1/10 change), to see if spells like Siphon Life or Corruption do indeed still use partial resists.
Maybe more opinions of people having played back then could tell us exactly what spell to look at.

**Comment by Skuzzlez** [comment]
> Your first video shows corruption ticks with various degrees of partial resistance, any idea why? https://www.youtube.com/watch?v=SRXUTgrZOvM&feature=youtu.be&t=16m55s

I've been reading more into this, apparently it was a bug that was introduced in Classic and isn't representative of vanilla.
https://us.forums.blizzard.com/en/wow/t/partial-resist-checks-on-dots/484801/15

**Comment by Skuzzlez** [comment]
Update: Came across this source from a 2008 guide (during TBC), credit to Scoper:

https://dwarfpriest.wordpress.com/2008/01/07/spell-hit-spell-penetration-and-resistances/

Probably worth reading the whole thing but this sticks out:

> DoTs like Vampiric Touch and Shadow Word: Pain are binary in their initial cast, but each damage tick can fall victim to resists just as a non-binary spell

**Comment by kelno** [comment]
I've been talking with some people and there doesn't seem to be a consensus that it should be resisting a lot on application.
I'm a bit perplexed about that sentence in dwarf priest blog because he mentions that then goes on the very specifics of effective resistance, which should be heavily impacted by this detail about dots, but he doesn't mention them again. But the post is overall very solid and he mentions something there so it could be good.
We'll do some additional data gathering there.

**Comment by Skuzzlez** [comment]
Thanks for looking into this again Kelno.

Here's are three separate guides from 2007 / 2008 each expressing the necessity to spec into Magical Resistances on hunter pets (coinciding with what Scoper mentioned above - that in comparison, resists feel relatively weak here).

https://blog.dnevnik.hr/wowfanblog/2008/07/1625114075/wow-arenahunterdruid-guide-za-2k-rating.html
> Get a Scorpion Pet. I can't stress it enough, it's critical. Give him full Nature resist, Full Frost Resist, Cobra reflexes, rank 5 Scorpid Poison, and the second or third highest Shadow Resist you can give.

https://wowguidelibrary.blogspot.com/2008/10/thersts-hunterdruid-2v2-guide-current-1.html
> I use a Scorpid pet for Scorpid poison, in order to cheese the dispel mechanic with Viper Stings. He has Avoidance, Cobra Reflexes, Frost Resistance (Rank 3), Great Stamina (Rank 3), Growl, Natural Armor (Rank 3), Nature Resistance (Rank 5), Scorpid Poison (Rank 5), and Shadow Resistance (Rank 5). Max out Shadow Resist to help it survive vs. lock, healer teams, and max Nature Resist to help resist Entangling Roots and Hibernates on it vs. druid, DPS teams.

https://customer121.blogspot.com/2007/10/2200-rated-2v2-hunter-strats-videos.html

> Get a SCORPION PET. Can't stress it enough, it's critical. Give him full Nature resist, Full Frost Resist, Cobra reflexes, rank 4 scorpid poison, and the second or third highest shadow resist you can give. 

And here's the big one:
https://gamefaqs.gamespot.com/pc/534914-world-of-warcraft/faqs/53491
Under the warlock section:

> your pet is one of the best ways to prevent enemy spellcasters from drinking.  Be aware that certain healers can kill pets over time (puppies can be wanded / meleed to death, vws can be dotted, etc)


Notice how he distinguishes that felhunters (with higher resist) should be wanded or even MELEE'd, whereas Voidwalkers can be dotted.

**Comment by kelno** [comment]
I see. Well I don't know what to make of those except that it says that spell resistance is worth having.

The last one I think still makes sense here, it seems to me he's talking about preventing for example priest/druids from drinking, and those even in the current implementation will have a very hard time dotting a high resistance target (~60% resistance, which also implies more than 25% full resist chance).

**Comment by Skuzzlez** [comment]
> even in the current implementation will have a very hard time dotting a high resistance target (~60% resistance, which also implies more than 25% full resist chance for a non binary).

The thing is, he even suggests wanding the felhunter over DoTing it. That is significant because wand damage is also affected by partial resistance. The only reason you'd ever wand over DoT a felhunter as a warlock is because of the additional binary resistance check.

**Comment by Jerrythemagicdog** [comment]
Videos found so far:
https://www.youtube.com/watch?v=yU6T3RA77ZE&feature=youtu.be&t=1377

https://youtu.be/ao6sbIKoh3o?t=628

https://youtu.be/ao6sbIKoh3o?t=1108

https://youtu.be/rEBL5Favl6Y?t=344

https://youtu.be/gJPt62uFiKU?t=1359

any more video evidence would be great

**Comment by kelno** [comment]
> The thing is, he even suggests wanding the felhunter over DoTing it. That is significant because wand damage is also affected by partial resistance. The only reason you'd ever wand over DoT a felhunter as a warlock is because of the additional binary resistance check on DoTs.
That's a good point

**Comment by Jerrythemagicdog** [comment]
> Be aware that certain healers can kill pets over time 
> 	(puppies can be wanded / meleed to death, vws can be dotted, etc)

I'm not sure what is this supposed to imply, of course those classes (like Priest) will have a harder time dotting a Felhunter than a Warlock (which is what we're primarily talking about) when they're running only 20 spell penetration in most cases. It's also not worth for those classes to continuously spam dot the pet and potentially mess up their regen by casting. They also do laughable damage to the pet with the little spell damage they have, low spell penetration and the 1 (or 2) DoT spells they have.

**Comment by Skuzzlez** [comment]
You make good points here. Perhaps it is likely that DoTs fully resist at 1/10th the rate of other binary spells, whereas their partial resists were, at the start of TBC, corrected to resist at the same rate as other non-binary spells, hence the dev thread. I'll continue to look for more info on this to see what I can find.

**Comment by Skuzzlez** [comment]
Ok I just found another major bug which might explain why Pets are getting shredded, at the very least by dots.

I tested Corruption on my pet with 0 resil, 140 shadow res. Max potential tick was 353 damage.

I then tested Corruption on my pet whilst equipping 362 resilience. The max tick again was 353 damage.

It looks like pets aren't inheriting resilience, at least against DoTs. I'll test it against other attacks now.

**Comment by Jerrythemagicdog** [comment]
Pets do not inherit resilience in TBC, that's a custom change we have tried on the PTR for some time at the early-WoTLK 40% value, players didn't want us to do that (Powerscoper you cite here among others) so we've reverted it.

**Comment by Skuzzlez** [comment]
My bad, thanks for clarifying this. 

Whilst testing resistances, I did notice that spells were getting resisted in percentages of 20%, i.e. 20%, 40%, 60% resisted rather than 25%, 50% - is there any particular reason behind this?

**Comment by Jerrythemagicdog** [comment]
Any custom changes are listed here: https://endless.gg/changes
This seems like a problem, we're looking into it.

**Comment by kelno** [comment]
Okay there was a real issue there, every time a spell should have been resisted 25%, it was resisted 20%, same for 50 -> 40, and 75 -> 60. This is now fixed and it makes spell resistance slightly stronger vs non binary spells.
Binary spells were not affected.

**Comment by kelno** [comment]
I found another issue this time making resistance slightly stronger for partial resists, but it's really convoluted. I've fixed it but I'm not good enough a mathematician to tell you how much it did affect it. My first intuition is that it was a very weak error in most cases but can become a bigger issue at high resistance scores. I'll get some help to validate this.
